File: frontend/pos-frontend/src/domain/entities/valueObjects/Money.ts
export class Money {
  private readonly cents: number;

  constructor(amount: number) {
    this.cents = Math.round(amount * 100);
  }

  get value(): number {
    return this.cents / 100;
  }

  add(other: Money): Money {
    return new Money((this.cents + other.cents) / 100);
  }

  subtract(other: Money): Money {
    return new Money((this.cents - other.cents) / 100);
  }

  multiply(factor: number): Money {
    return new Money((this.cents * factor) / 100);
  }

  isGreaterThan(other: Money): boolean {
    return this.cents > other.cents;
  }

  isLessThan(other: Money): boolean {
    return this.cents < other.cents;
  }

  equals(other: Money): boolean {
    return this.cents === other.cents;
  }

  toString(): string {
    return `$${this.value.toFixed(2)}`;
  }
}

File: backend/pos-backend/src/main/java/com/pos/backend/domain/model/valueobject/Money.java
package com.pos.backend.domain.model.valueobject;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

/**
 * Value object que representa un valor monetario.
 * Siempre 2 decimales, no negativo.
 */
public record Money(BigDecimal amount) {

    public static final Money ZERO = new Money(BigDecimal.ZERO);
    private static final int SCALE = 2;
    private static final RoundingMode ROUNDING = RoundingMode.HALF_UP;

    public Money {
        Objects.requireNonNull(amount, "Amount cannot be null");
        if (amount.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Money amount cannot be negative: " + amount);
        }
        amount = amount.setScale(SCALE, ROUNDING);
    }

    public static Money of(BigDecimal value) {
        return new Money(value);
    }

    public static Money of(double value) {
        return new Money(BigDecimal.valueOf(value));
    }

    public static Money of(String value) {
        return new Money(new BigDecimal(value));
    }

    public Money add(Money other) {
        return new Money(this.amount.add(other.amount));
    }

    public Money subtract(Money other) {
        BigDecimal result = this.amount.subtract(other.amount);
        if (result.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Subtraction would result in negative money");
        }
        return new Money(result);
    }

    public Money multiply(int factor) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(factor)));
    }

    public Money multiply(BigDecimal factor) {
        return new Money(this.amount.multiply(factor));
    }

    /** Calcula un porcentaje del valor (ej: percentage(19) = 19% del monto) */
    public Money percentage(int percent) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(percent))
                                    .divide(BigDecimal.valueOf(100), SCALE, ROUNDING));
    }

    public boolean isGreaterThan(Money other) {
        return this.amount.compareTo(other.amount) > 0;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return this.amount.compareTo(other.amount) >= 0;
    }

    public boolean isZero() {
        return this.amount.compareTo(BigDecimal.ZERO) == 0;
    }

    @Override
    public String toString() {
        return amount.toPlainString();
    }
}
